src/DynamoCoreWpf/Views/Core/NodeView.xaml.cs

```diff
@@ -687,26 +688,537 @@ internal void TogglePreviewControlAllowance()
 
         #endregion
 
-        private void OptionsButton_Click(object sender, RoutedEventArgs e)
+        /// <summary>
+        /// A static class containing methods for dynamically building a Node's Context Menu.
+        /// </summary>
+        internal class NodeContextMenuBuilder
         {
-            Guid nodeGuid = ViewModel.NodeModel.GUID;
-            ViewModel.DynamoViewModel.ExecuteCommand(
-                new DynCmd.SelectModelCommand(nodeGuid, Keyboard.Modifiers.AsDynamoType()));
+            // A reference to the NodeViewModel whose ContextMenu is being built.
+            private NodeViewModel NodeViewModel { get; }
+            private ContextMenu ContextMenu { get; }
 
-            viewModel.OnSelected(this, EventArgs.Empty);
-            grid.ContextMenu.DataContext = viewModel;
-            grid.ContextMenu.IsOpen = true;
+            // Constructor
+            internal NodeContextMenuBuilder(NodeViewModel nodeViewModel, ContextMenu contextMenu)
+            {
+                NodeViewModel = nodeViewModel;
+                ContextMenu = contextMenu;
+            }
+
+            /// <summary>
+            /// Adds or inserts items to a context menu.
+            /// </summary>
+            /// <param name="menuItem"></param>
+            internal void AddContextMenuItem(MenuItem menuItem) => ContextMenu.Items.Add(menuItem);
```

Comment:
If you change the signature of these methods, you could make them into a nice fluid builder 

(dunno type of ContextMenu below)

```suggestion
            internal ContextMenuType AddContextMenuItem(MenuItem menuItem){
            ContextMenu.Items.Add(menuItem);
            return this;
            }
```

so usage would be like this (notice the extra conditional fluid method too)

```csharp
this.ContextMenu
    .AddContextMenuItem(NodeContextMenuBuilder.Whatever())
    .AddContextMenuItem(NodeContextMenuBuilder.WhateverElse())
    .AddContextMenuSeparator(NodeContextMenuBuilder.XXX())
    .ConditionalAddContextMenuItem(NodeContextMenuBuilder.IsInputOutputMenuItem(), NodeViewModel.IsInput || NodeViewModel.IsOutput)
```

Reply:
could also be `extension methods` and move them to a separate `NodeViewExtensions` file - but dunno if that fits this codebase.

Reply:
Have split out Add and Build methods.